contributors/design-proposals/node/troubleshoot-running-pods.md

```diff
@@ -192,96 +225,100 @@ type EphemeralContainer struct {
+`EphemeralContainer.Status` introduces no new data, but the kubelet must also
+now populate `EphemeralContainer.Spec` &
+`EphemeralContainer.TargetContainerName`.
+
+The kubelet already persists container metadata as CRI
+[labels](https://github.com/kubernetes/kubernetes/blob/v1.10.0-alpha.0/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto#L606)
+and
+[annotations](https://github.com/kubernetes/kubernetes/blob/v1.10.0-alpha.0/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto#L613).
+The entire v1.Container used in the request will be serialized and stored as a
+runtime annotation. The value of `TargetContainerName` will be stored as a
+runtime label. Persisting this data in the runtime means it will survive kubelet
+restarts.
+
+At least for the Docker runtime, this is [an intended use of docker
+labels](https://docs.docker.com/engine/userguide/labels-custom-metadata/#value-guidelines).
+Docker does not document the maximum length of labels in its API. Empirically,
+it supports up to the 64K constraint of the docker client's `bufio.Scanner`
+size. Because the container spec may be examined in security sensitive contexts
+like admission control, we will conservatively limit the size of the spec to 32K
+and add a 32K minimum label length test to runtime qualification.
+
+`EphemeralContainer` is populated by the kubelet in the same way as regular
+container statuses. This is sent to the API server and displayed by `kubectl
+describe pod`.
 
 ### Creating Debug Containers
 
-1.  `kubectl` invokes the exec API as described in the preceding section.
+1.  `kubectl` invokes the new API as described in the preceding section.
 1.  The API server checks for name collisions with existing containers, performs
-    admission control and proxies the connection to the kubelet's
-    `/exec/$NS/$POD_NAME/$CONTAINER_NAME` endpoint.
-1.  The kubelet instructs the Runtime Manager to create a Debug Container.
-1.  The runtime manager uses the existing `startContainer()` method to create a
-    container in an existing pod. `startContainer()` has one modification for
-    Debug Containers: it creates a new runtime label (e.g. a docker label) that
-    identifies this container as a Debug Container.
-1.  After creating the container, the kubelet schedules an asynchronous update
-    of `PodStatus`. The update publishes the debug container status to the API
-    server at which point the Debug Container becomes visible via `kubectl
-    describe pod`.
-1.  The kubelet will upgrade the connection to streaming and attach to the
-    container's console.
-
-Rather than performing the implicit attach the kubelet could return success to
-the client and require the client to perform an explicit attach, but the
-implicit attach maintains consistent semantics across `/exec` rather than
-varying behavior based on parameters.
+    validation, admission control and proxies the connection to the kubelet's
+    `/ephemeralContainers/$NS/$POD_NAME` endpoint.
```

Comment:
it would be helpful to clarify at what point this call completes (e.g. does it block on the syncPod call or return immediately)

Reply:
nm, I see the details below